src/metabase/models/spec_update.clj

```diff
@@ -0,0 +1,172 @@
+(ns metabase.models.spec-update
+  "Perform Creation/Update/Deletion with a spec."
+  (:require
+   [clojure.string :as str]
+   [metabase.util :as u]
+   [metabase.util.log :as log]
+   [metabase.util.malli :as mu]
+   [toucan2.core :as t2]))
+
+(def ^{:dynamic true
+       :private true} *current-path* [])
+
+(defn- current-path
+  []
+  (str (str/join #" / " (map #(if (keyword? %) (name %) %) *current-path*)) ":"))
+
+;; TODO: add :id-column, currently it assumes all models use :id
+(def ^:private Spec
+  [:schema {:registry {::spec [:map {:closed true}
+                               [:model                         [:fn #(isa? % :metabase/model)]]
+                               ;; whether this nested model is a sequentials with respect to the parent model
+                               [:multi-row?   {:optional true} :boolean]
+                               ;; the foreign key column in the nested model with respect to the parent model
+                               [:fk-column    {:optional true} :keyword]
+                               ;; a function to get row into a comparable state
+                               [:compare-row  {:optional true} :any #_[:=> [:cat :map] :map]]
+                               [:nested-specs {:optional true} [:map-of :keyword [:ref ::spec]]]]}}
+   [:ref ::spec]])
+
+(defmacro ^:private with-enter-path
+  [path & body]
+  `(binding [*current-path* (conj *current-path* ~path)]
```

Comment:
It doesn't seem like the code would be that much trickier if it just worked with a path variable. I'd much prefer not using dynamic state.

Reply:
Huge +1, just make this an explicit argument and it'll be easier to debug too.